File: Java-deprecated/com.dexaint.dukechess/src/main/java/com/dexaint/dukechess/movement/Command.java
package com.dexaint.dukechess.movement;

import org.apache.commons.lang3.ArrayUtils;

import com.dexaint.dukechess.flow.Field;
import com.dexaint.dukechess.flow.Player;

public class Command extends MovementImpl {
	public int[] validateMovement(Field field, int pos, Destination d, Player p) {
		int[] ret = new int[]{};
		
		int[] offsets = d.dest2Offset(p);
		
		if (!isInField(field, pos, offsets)) return ret; // Short-Circuit-OR
		
		ret = ArrayUtils.add(ret, offset2Dest(field, pos, d.dest2Offset(p)));
		
		return ret;
	}
}

File: dukechess/js-backend/movement/Command.js
const MovementImpl = require("./MovementImpl");
const DestUtils = require("./DestUtils");

class Command extends MovementImpl {
	validateMovement(field, pos, d, p) {
		var ret = [];
		
		var offsets = DestUtils.dest2Offset(d, p);
		
		if (!this.isInField(field, pos, offsets)) return ret;
		
		ret = ret.concat(this.offset2Dest(field, pos, offsets));
		
		return ret;
	}

	validateControlArea(field, pos, d, p) {
		var ret = [];

		var offsets = DestUtils.dest2Offset(d, p);

		if (!this.isInField(field, pos, offsets)) return ret;

		if (this.hasMyChess(field, pos, offsets, p)) return ret;

		ret = ret.concat(this.offset2Dest(field, pos, offsets));

		return ret;
	}
}

module.exports = Command;
